Sprint 16 mvvm

superaydumov opened the pull request:
Pull request according to Yandex Praktikum course (sprint_16).

superaydumov committed:
Added GeometricParams structure.
Refactored TrackersViewController for TrackerCollectionViewCell to be dynamically calculated depending on screen size.

superaydumov committed:
Little changes in code.

superaydumov committed:
Added OnboardingViewController class to show onboarding screens when the app is open.
Refactored SceneDelegate class to show OnboardingViewController first.
Added UIView+Extension to add picture as background image to viewController's view.

superaydumov committed:
Refactored OnboardingViewController and SceneDelegate classes to show onboarding only when the app opens for the first time.

superaydumov committed:
CategoryCreationViewController class war created.
Refactored CategoryViewController to jump to CategoryCreationViewController when button is tapped.

superaydumov committed:
Refactored CategoryViewController: tableView for categories and loading categories from trackerCoreData were added.
Refactored EventCreatorViewController to show selected category from CategoryViewController:
- add CategoryViewControllerDelegate protocol, categoryButtonSubtitle property, categorySubtitleUpdate and createCategory methods;
- also little changes in class.

superaydumov committed:
Refactored TrackersViewController to use performBatchUpdates instead of reloadData in collectionView.
Updated deleTrackerRecord method in TrackerRecordCoreData class (deleting trackerRecord worked incorrect).
Refactored CategoryViewController to show separatorInset between cells correctly.

superaydumov committed:
Fixed issue with CategoryCreationViewController (git problem).

superaydumov committed:
Fixed issue with separator between cells in CategoryViewController's tableView (it was hidden when user added new category).

superaydumov committed:
Refactored CategoryViewController for tableView to show separator between cells correctly after appending new category.

superaydumov committed:
Added methods to TrackerCategoryStore class to update and delete existing category.
Modified CategoryCreatorViewController to show itself when user creates and modifies category.
Added contextMenuConfiguration method in CategoryViewController to show context menu with long press on tableView's cell.

superaydumov committed:
Added alert when user tries to delete category in CategoryViewController.
Also added observer in TrackersViewController to call updateCategories method when category is deleted.

superaydumov committed:
Refactored CategoryViewController, EventCreatorViewController and CategoryCreatorViewController classes to add MVVM pattern logic.
Added CategoryViewModel, CategoryCreatorViewControllerDelegate and CategoryViewModelDelegate.

andrew-kuzn-yndx commented on Tracker/OnboardingViewController.swift:
```diff
@@ -0,0 +1,178 @@
+    
+    private lazy var blueViewController: UIViewController = {
+        let bluePage = UIViewController()
+        let blueImage = "blueBackground.png"
```
Хорошей практикой является вывод строковых значений в какую нибудь внутреннюю сттруктуру или extension - чтобы они хранились отдельно) Например, прям в viewcontroller можно сделать так:

enum Constants {
static let blueViewControllerImage: String = "blueBackground.png"
}

а тут уже вызывать через Constants. blueViewControllerImage

superaydumov replied:
Done

andrew-kuzn-yndx commented on Tracker/OnboardingViewController.swift:
```diff
@@ -0,0 +1,178 @@
+    
+    private lazy var button: UIButton = {
+        let button = UIButton(type: .system)
+        button.setTitle("Вот это технологии!", for: .normal)
```
Такие строки тоже можно выводить в Constants

superaydumov replied:
Done

andrew-kuzn-yndx commented on Tracker/Trackers/EventCreatorViewController/Category/CategoryViewController.swift:
```diff
@@ -113,10 +154,116 @@ final class CategoryViewController: UIViewController {
+        } else if indexPath.row == viewModel.categories.startIndex {
+            cell.layer.cornerRadius = 16
+            cell.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
+            cell.separatorInset = UIEdgeInsets(top: 0, left: 16, bottom: 0, right: 16)
```
Лучше сделать вьюшки ячейки как private, и менять их внутри ячейки.
Работать напрямую со свойствами ячейки из другого класса - не лучшая практика.  Лучше сделать в ячейке публичную функцию, например configCell с аргументами на вход, и уже в ней менять все сввойства.
Всё таки состояние ячейки - это её ответственность, а не контроллера)

superaydumov replied:
Done

andrew-kuzn-yndx commented on Tracker/Trackers/TrackersViewController.swift:
```diff
@@ -139,6 +140,9 @@ final class TrackersViewController: UIViewController {
         trackerCategoryStore.delegate = self
+        
+        NotificationCenter.default.addObserver(self, selector: #selector(reloadCollectionView(sender:)), name: NSNotification.Name("reloadCollectionView"), object: nil)
+
```
Всегда помни, что при добавлении Observer, нужно предусматривать и его удаление, чтобы не создать утечку памяти. Стоит добавить удаление observer

superaydumov replied:
Done

superaydumov committed:
Refactored viewControllers after reviewer's comments:
- added Keys structures to combine all string constants in one place;
- refactored viewControllers with cells to exclude using cell parameters in cellForRowAt/cellForItemAt methods (now cell view's are private);
- added methods to get data from cells where needed;
- added observer post in CategoryCreatorViewController to update collectionView in TrackersViewController when category name is changed;
- added deinit in TrackersViewController to delete observer.
